File: Analysis/Ex/LoKiExample/python/LoKiExample/LoKi_MCMuon.py
#!/usr/bin/env gaudirun.py
# =============================================================================
# $Id: LoKi_MCMuon.py,v 1.3 2009-10-15 09:28:59 pkoppenb Exp $ 
# =============================================================================
## @file
#  The configuration file to run LoKi_MCMuon example
#
#   This file is a part of LoKi project - 
#     "C++ ToolKit  for Smart and Friendly Physics Analysis"
#
#  The package has been designed with the kind help from
#  Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
#  contributions and advices from G.Raven, J.van Tilburg, 
#  A.Golutvin, P.Koppenburg have been used in the design.
#
#  By usage of this code one clearly states the disagreement 
#  with the campain of Dr.O.Callot et al.: 
#  "No Vanya's lines are allowed in LHCb/Gaudi software."
#
#  @author Vanya BELYAEV Ivan.Belyaev@nikhef.nl
#  @date   2008-06-07
# =============================================================================
"""
Configuration file for LoKiExample package

This file is a part of LoKi project - 
\"C++ ToolKit  for Smart and Friendly Physics Analysis\"

The package has been designed with the kind help from
Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
contributions and advices from G.Raven, J.van Tilburg, 
A.Golutvin, P.Koppenburg have been used in the design.

By usage of this code one clearly states the disagreement 
with the campain of Dr.O.Callot et al.: 
\"No Vanya's lines are allowed in LHCb/Gaudi software.\"

"""
# =============================================================================
__author__  = " Vanya BELYAEV Ivan.Belyaev@nikhef.nl "
__version__ = " CVS Tag $Name: not supported by cvs2svn $, version $Revision: 1.3 $ "
# =============================================================================
from  Gaudi.Configuration import * 


## create & configure own algorithm:
from  Configurables import LoKi__MCMuons as MCMuons
alg = MCMuons (
    "MCMuons"                                           , ## the instance name 
    InputLocations = [ 'StdTightMuons']                 , ##   input particles 
    PP2MCs         = [ 'Relations/Rec/ProtoP/Charged' ] , ##   PP -> MC tables
    NTupleLUN      = "MCMU"
    )

## get input data:
from LoKiExample.Bs2Jpsiphi_mm_data import Inputs as INPUT 

## confgure the application itself:
from  Configurables import DaVinci 
DaVinci (
    DataType       = 'DC06'  , ## Data type  
    Simulation     = True    , ## Monte Carlo 
    Hlt            = False      ,
    #
    UserAlgorithms = [ alg ] , ## let DaVinci know about local algorithm
    # delegate this properties to Event Selector 
    EvtMax        = 500      ,  
    SkipEvents    = 0        ,
    Input         = INPUT    , ## the list of input data files
    # delegate to Histogram Persistency Service
    HistogramFile = "PsiPhi_Histos.root"
    )

## n-Tuples
from Configurables import NTupleSvc
svc = NTupleSvc() 
svc.Output += [ "MCMU DATAFILE='MCMuon_Tuples.root' TYP='ROOT' OPT='NEW'" ] 


# =============================================================================
# The END
# =============================================================================

File: Analysis/Ex/LoKiExample/src/LoKi_MCMuon.cpp
// $Id: LoKi_MCMuon.cpp,v 1.10 2008-07-09 17:00:48 ibelyaev Exp $
// ============================================================================
// Include files 
// ============================================================================
// LoKi
// ============================================================================
#include "LoKi/AlgoMC.h"
#include "LoKi/ParticleCuts.h"
#include "LoKi/MCParticleCuts.h"
#include "LoKi/PhysMCParticleCuts.h"
// ============================================================================
/** @file
 *
 *  This file is a part of LoKi project - 
 *    "C++ ToolKit  for Smart and Friendly Physics Analysis"
 *
 *  The package has been designed with the kind help from
 *  Galina PAKHLOVA and Sergey BARSUK.  Many bright ideas, 
 *  contributions and advices from G.Raven, J.van Tilburg, 
 *  A.Golutvin, P.Koppenburg have been used in the design.
 *
 *  @author Vanya BELYAEV ibelyaev@physics.syr.edu
 *  @date   2003-03-27 
 */
// ============================================================================
LOKI_MCALGORITHM( MCMuons ) 
{
  using namespace LoKi::Cuts ;
  using namespace LoKi::Types ;
  
  // get all reconstructed muons
  Range   muons   = select   ( "Muons" , ABSID == "mu+" ) ;
  
  // get all MC muons 
  MCRange mcmuons = mcselect ( "MCmu" , MCABSID == "mu+" ) ;
  
  // create the matching  object 
  MCMatch mc = mcTruth();
  
  // check for matching with ALL muons 
  Cut mccut1 = MCTRUTH( mc , mcmuons.begin() , mcmuons.end() ) ;
  
  // check for matching with muon with highest PT 
  MCRange::iterator imu = 
    select_max ( mcmuons.begin() , mcmuons.end() , MCPT ) ;

  Cut mccut2 = ( mcmuons.end() == imu ) ? 
    Cut( MCTRUTH( mc , *imu ) ) : Cut( PNONE ) ;
  
  Tuple tuple = nTuple ( "Muons" ) ;
  
  const double GeV = Gaudi::Units::GeV ;
  
  // loop over all recontructed muons 
  for ( Range::iterator imuon = muons.begin() ; 
        muons.end() != imuon ; ++imuon ) 
  {
    const LHCb::Particle* muon = *imuon ;
    if( 0 == muon ) { continue ; }
    // ID variable 
    tuple -> column ( "DLLmu" , PIDmu  ( muon )  / GeV ) ;
    tuple -> column ( "PT"    , PT     ( muon )  / GeV ) ;
    tuple -> column ( "RCM"   , muon->momentum() / GeV ) ; //4-momentum 
    tuple -> column ( "mc1"   , mccut1 ( muon )  ) ; // true muon?
    tuple -> column ( "mc2"   , mccut2 ( muon )  ) ; // true muon with highest PT
    
    // find MC muon which 'matches' with the recontructed muon
    MCRange::iterator im = 
      mc->match ( muon , mcmuons.begin() , mcmuons.end() ) ;
    
    // 4-momentum of MC muon
    LoKi::LorentzVector lv ;
    
    if( mcmuons.end() != im &&  0 != *im ) 
    { lv    = (*im)->momentum() / GeV ; }
    
    tuple -> column ( "MCM" , lv    ).ignore () ; 
    
    tuple -> write  () .ignore()  ;
  }
  //
  return StatusCode::SUCCESS ;  
} 
// ============================================================================
// The END 
// ============================================================================
